[dist_optim] add distributed functional Adadelta optimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50630 [dist_optim] add warning to distributed optimizer
* #50620 [dist_optim] add distributed functional AdamW optimizer
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* **#50623 [dist_optim] add distributed functional Adadelta optimizer**
* #50618 [dist_optim] add distributed functional sgd optimizer
* #50624 [dist_optim] add distributed functional Adam optimizer

Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

Differential Revision: [D25932772](https://our.internmc.facebook.com/intern/diff/D25932772)

wanchaol committed:
[dist_optim] add distributed functional Adadelta optimizer

Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adadelta optimizer"


Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

[ghstack-poisoned]

rohan-varma reviewed (commented):
Looks like there is a potentially related CI failure, could you confirm?

https://app.circleci.com/pipelines/github/pytorch/pytorch/260823/workflows/c7ded86c-e48d-4482-a8eb-a53c54a412a9/jobs/10202460

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 422, in wrapper
    fn(*args, **kwargs)
  File "test_optim.py", line 380, in test_multi_tensor_optimizers
    self.assertEqual(p1, p2)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1179, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 6 element(s) (out of 6) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.14733898893893732 (0.020583703180005117 vs. 0.16792269211894245), which occurred at index (2, 1).
		
```

wanchaol committed:
Update on "[dist_optim] add distributed functional Adadelta optimizer"


Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

Differential Revision: [D25932772](https://our.internmc.facebook.com/intern/diff/D25932772)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adadelta optimizer"


Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

Differential Revision: [D25932772](https://our.internmc.facebook.com/intern/diff/D25932772)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional Adadelta optimizer"


Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

Differential Revision: [D25932772](https://our.internmc.facebook.com/intern/diff/D25932772)

[ghstack-poisoned]

rohan-varma reviewed (approved):
CI looks good now, LGTM

wanchaol committed:
Update on "[dist_optim] add distributed functional Adadelta optimizer"


Add TorchScript compatible Adadelta functional optimizer to distributed optimizer

Differential Revision: [D25932772](https://our.internmc.facebook.com/intern/diff/D25932772)

[ghstack-poisoned]
